system/jhtree/jhinplace.cpp

```diff
@@ -0,0 +1,2681 @@
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "platform.h"
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#ifdef __linux__
+#include <alloca.h>
+#endif
+#include <algorithm>
+
+#include "jmisc.hpp"
+#include "jset.hpp"
+#include "hlzw.h"
+
+#include "ctfile.hpp"
+#include "jhinplace.hpp"
+#include "jstats.h"
+
+#ifdef _DEBUG
+//#define SANITY_CHECK_INPLACE_BUILDER     // painfully expensive consistency check
+//#define TRACE_BUILDING
+#endif
+
+static constexpr size32_t minRepeatCount = 2;       // minimum number of times a 0x00 or 0x20 is repeated to generate a special opcode
+static constexpr size32_t minRepeatXCount = 3;      // minimum number of times a byte is repeated to generate a special opcode
+static constexpr size32_t maxQuotedCount = 31 + 256; // maximum number of characters that can be quoted in a row
+static constexpr size32_t repeatDelta = (minRepeatCount - 1);
+static constexpr size32_t repeatXDelta = (minRepeatXCount - 1);
+
+constexpr size32_t getMinRepeatCount(byte value)
+{
+    return (value == 0x00 || value == 0x20) ? minRepeatCount : minRepeatXCount;
+}
+
+enum SquashOp : byte
+{
+    SqZero      = 0x00,           // repeated zeros
+    SqSpace     = 0x01,           // repeated spaces
+    SqQuote     = 0x02,           // Quoted text
+    SqRepeat    = 0x03,           // any character repeated
+    SqOption    = 0x04,           // an set of option bytes
+    SqSingle    = 0x05,           // a single character - count indicates which of the common values it is.
+    SqSingleX   = 0x06,           // A SqSingle followed by a single byte
+
+    //What other options are worth doing?
+    //SqXSingle   = 0x07,         // A single byte followed by a SqSingle?  Maybe 2.5% on my sample index.
+    //SqMatchX                    // Matches a string at offset (cur-1-count) - delta8/16.  Only likely to help with repeated trailing key segments
+};
+
+//inline unsigned reverseBytes(unsigned value) { return __builtin_bswap32(value); }
+
+//MORE: What other special values are worth encoding in real life indexes?
+static constexpr byte specialSingleChars[31] = {
+    0, 1, 2, 3, 4, 5, 6, 7,                 // 1-8      Useful for small numbers
+    ' ', '0', 'F', 'M', 'U',                // 9-13     Single padding and gender
+                                            // 14-31    Currently unused.
```

Comment:
Other digits would seem likely to occur?

Reply:
This is something that would be good to tune on some real-life indexes.